### PR TITLE
feat(cuda): add C-level simple API mirroring CPU one-shot wrappers

### PR DESCRIPTION
**Pull request by DiamonDinoia**
Adds 36 entry points (3 dims x 3 types x {single,many} x {single,double}) that wrap the cuFINUFFT plan -> setpts -> execute -> destroy sequence into a single call. Pointers are device pointers; argument order matches the CPU finufft1d1 family in src/c_interface.cpp. The plan is held in a unique_ptr so error paths still release GPU resources.

A cross-check test exercises every (dim, type, precision) combination against the 4-step plan path, and the C ABI is smoke-tested from test/cuda/public_api_test.c. The new surface is documented in docs/c_gpu.rst.

Reference: the prior cufinufft simple-interface draft by @janden on janden/update_cu_api

### CODE REVIEW

**Comment by mreineck** [comment]
Good idea!

**Comment by ahbarnett** [comment]
@janden we would love your input on this, as a GPU user! (it doesn't affect Py wrappers, though...).

File: docs/c_gpu.rst [hunk 1]
```diff
@@ -19,6 +19,67 @@ You will also want to read the examples in ``examples/cuda`` and ``test/cuda/cuf
...
+Returns ``0`` on success, otherwise an error code (see ``finufft_errors.h``).
+
+A complete example using the simple interface in place of the 4-step calls
+from the :ref:`Getting started <c_gpu>` walkthrough below is:
+
+.. code-block:: c
+
+    cufinufftf1d1(M, d_x, d_c, 1, 1e-6f, N, d_f, NULL);
+
+This single call replaces the ``makeplan``/``setpts``/``execute``/``destroy``
```
Comment (ahbarnett) [inline docs/c_gpu.rst]: replaces -> is equivalent to the ... sequence required with the plan interface.

File: docs/c_gpu.rst [hunk 2]
```diff
@@ -19,6 +19,67 @@ You will also want to read the examples in ``examples/cuda`` and ``test/cuda/cuf
...
+sequence. Type-3 calls also take frequency-target arrays (``s`` for 1D,
+``s,t`` for 2D, ``s,t,u`` for 3D), exactly as in the CPU API. The ``many``
+variants take an ``n_transf`` parameter as the first argument and use the
+same set of NU points for ``n_transf`` consecutive transforms (with input
+and output arrays sized accordingly).
+
+The simple interface is intended for cases where the plan would otherwise be
+used exactly once. When the same NU points are reused across several
+transforms, use the 4-step plan interface below for better performance.
```
Comment (ahbarnett) [inline docs/c_gpu.rst]: nice explanation

File: test/cuda/cufinufft_simple_test.cu [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+  if (worst > rtol) {
+    std::fprintf(stderr, "  worst rel error = %g (tol %g)\n", double(worst),
+                 double(rtol));
+    return false;
+  }
+  return true;
+}
+
+template<typename T> struct Fixture {
+  using cuc = typename Traits<T>::cuc;
```
Comment (ahbarnett) [inline test/cuda/cufinufft_simple_test.cu:81]: I don't know Traits and don't know what this Fixture object is for - could you add a comment?

File: test/cuda/cufinufft_simple_test.cu [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+template<typename T> Fixture<T> make_fixture(int dim, int type, std::mt19937 &rng) {
+  Fixture<T> f;
+  f.dim = dim;
+  f.type = type;
+  f.M = 137;
+  f.N1 = 16;
+  f.N2 = (dim >= 2) ? 12 : 1;
+  f.N3 = (dim >= 3) ? 8 : 1;
+  f.Nk = 23;
+  std::uniform_real_distribution<T> uni(T(-3.14), T(3.14));
```
Comment (ahbarnett) [inline test/cuda/cufinufft_simple_test.cu:107]: Is 3.14 supposed to be M_PI ?  (otherwise it might miss edge cases where NU pts right on the domain edge).

File: test/cuda/cufinufft_simple_test.cu [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+  std::uniform_real_distribution<T> uni_s(T(-2), T(2));
+  std::uniform_real_distribution<T> uni_c(T(-1), T(1));
```
Comment (ahbarnett) [inline test/cuda/cufinufft_simple_test.cu]: ok, uni_c is used as Re,Im parts of complex coeffs, fine.   But what is the rationale of uni_s? Why is the range [-2,2]? Add comment?

File: test/cuda/cufinufft_simple_test.cu [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+  } else {
+    cudaMemcpy(d_fk, f.fk_in.data(), f.fk_count() * sizeof(cuc), cudaMemcpyHostToDevice);
+  }
+
+  const int64_t nmodes[3] = {f.N1, f.N2, f.N3};
+  const int Nk32 = (f.type == 3) ? static_cast<int>(f.Nk) : 0;
+  std::vector<std::complex<T>> out;
+  if constexpr (std::is_same_v<T, float>) {
+    cufinufftf_plan plan{};
+    CHECK_RET(cufinufftf_makeplan(f.type, f.dim, nmodes, 1, 1, T(1e-5), &plan, nullptr),
```
Comment (ahbarnett) [inline test/cuda/cufinufft_simple_test.cu]: The value 1e-5 or similar appears scattered across this test code in several places. I'm guessing that sometimes it is a requested tol, and sometimes a check for the relative error (eg 5e-5).  Could you follow our previous math-checking codes and give these informative names that are defined in one place only?

File: test/cuda/cufinufft_simple_test.cu [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+template<typename T> int compare_one(int dim, int type, std::mt19937 &rng) {
+  auto f = make_fixture<T>(dim, type, rng);
+  auto via_p = run_plan(f);
+  auto via_s = run_simple(f);
+  if (via_p.empty() || via_s.empty()) return 1;
+  // Tolerance: the algorithm is the same; only atomic-add non-determinism
+  // separates the two runs. Float type-3 chains spreading + FFT + interpolation
+  // so the error accumulates more; allow a looser bound there.
+  const T rtol = std::is_same_v<T, float> ? T(5e-3) : T(1e-9);
```
Comment (ahbarnett) [inline test/cuda/cufinufft_simple_test.cu]: see above

File: test/cuda/cufinufft_simple_test.cu [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+  d_c = dev_alloc<cuc>(f.c_count());
+  d_fk = dev_alloc<cuc>(f.fk_count());
+
+  if (f.type == 1 || f.type == 3) {
+    cudaMemcpy(d_c, f.c_in.data(), f.c_count() * sizeof(cuc), cudaMemcpyHostToDevice);
+  } else {
+    cudaMemcpy(d_fk, f.fk_in.data(), f.fk_count() * sizeof(cuc), cudaMemcpyHostToDevice);
+  }
+
+  const T eps = std::is_same_v<T, float> ? T(1e-5) : T(1e-9);
```
Comment (ahbarnett) [inline test/cuda/cufinufft_simple_test.cu]: see above

File: test/cuda/cufinufft_simple_test.cu [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+  return rc;
+}
+
+} // namespace
+
+int main() {
+  int rc = 0;
+  rc |= run_all<float>();
+  rc |= run_all<double>();
+  return rc;
```
Comment (ahbarnett) [inline test/cuda/cufinufft_simple_test.cu:361]: good

File: test/cuda/public_api_test.c [hunk 10]
```diff
@@ -182,11 +182,59 @@ int test_double(int M, int N) {
   return 0;
 }
 
+// Smoke test for the C-level simple API: same 1D type-1 problem as test_float,
+// but using cufinufftf1d1 (one call) instead of the 4-step plan path. Verifies
+// the C ABI (this TU is compiled as C, not C++).
```
Comment (ahbarnett) [inline test/cuda/public_api_test.c:187]: add comment that no math test

**Review by ahbarnett — CHANGES_REQUESTED** [review]
I checked it all built and ctest ok on RTX A6000, on sm86.
I like the new cufinufft_simple_test.cu  but gave some improvements in the style of existing tests, for maintainability.
Thanks!

File: test/cuda/cufinufft_simple_test.cu (hunk 4 above)
Comment (DiamonDinoia) [reply test/cuda/cufinufft_simple_test.cu:107]: we don't care about this here, the tests just have to check that the correct functions are called here.

File: test/cuda/public_api_test.c (hunk 10 above)
Comment (DiamonDinoia) [reply test/cuda/public_api_test.c:187]: smoke test implies it since by definition a smoke test only check that calling the function does not crash

File: test/cuda/cufinufft_simple_test.cu (hunk 6 above)
Comment (ahbarnett) [reply test/cuda/cufinufft_simple_test.cu]: Same applies to 1e-9 for double tol, etc.

**Comment by DiamonDinoia** [comment]
I think I did it. Did I miss something?

**Review by ahbarnett — APPROVED** [review]
Great. I didn't see the tol consts until now - looks good.

File: src/cuda/c_interface.cpp [hunk 11]
```diff
@@ -170,4 +172,296 @@ void cufinufft_default_opts(cufinufft_opts *opts)
...
+  cufinufft_opts planopts;
+  if (popts)
+    planopts = *popts;
+  else
+    cufinufft_default_opts(&planopts);
+
+  auto plan = std::make_unique<cufinufft_plan_t<T>>(type, n_dims, nmodes32, iflag,
+                                                    n_transf, eps, planopts);
+  const int warn = plan->eps_too_small ? FINUFFT_WARN_EPS_TOO_SMALL : 0;
+  plan->setpts((int)nj, xyz[0], xyz[1], xyz[2], (int)nk, stu[0], stu[1], stu[2]);
```
Comment (janden) [inline src/cuda/c_interface.cpp:216]: Should we not be catching errors from setpts and execute as well?

File: src/cuda/c_interface.cpp [hunk 12]
```diff
@@ -170,4 +172,296 @@ void cufinufft_default_opts(cufinufft_opts *opts)
...
+  plan->execute(cj, fk);
+  return warn;
+}
+
+template<typename T>
+int guru13(int n_dims, int type, int n_transf, i64 nj,
```
Comment (janden) [inline src/cuda/c_interface.cpp:225]: Could you comment briefly about why these are needed? To make sure const qualifiers are respected?

File: src/cuda/c_interface.cpp [hunk 13]
```diff
@@ -170,4 +172,296 @@ void cufinufft_default_opts(cufinufft_opts *opts)
...
+           const std::array<const T *, 3> &xyz, const cuda_complex<T> *cj, int iflag,
+           T eps, const std::array<ci64, 3> &n_modes, i64 nk,
+           const std::array<const T *, 3> &stu, cuda_complex<T> *fk,
+           const cufinufft_opts *popts) {
+  return safe_finufft_call([&]() -> int {
```
Comment (janden) [inline src/cuda/c_interface.cpp:230]: What is `safe_finufft_call` and why is it needed?

File: test/cuda/cufinufft_simple_test.cu [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+  cudaMemcpy(v.data(), src, n * sizeof(T), cudaMemcpyDeviceToHost);
+  return v;
+}
+
+// Approximate equality. GPU spreading uses atomicAdd which is not bit-
+// deterministic for float, so two independent runs of the same type-1 / type-3
+// transform can differ by a few ULPs even though the algorithm is the same.
+// Type-2 paths (interpolation only, no atomics) compare bit-identically; we
+// allow a small relative tolerance uniformly to keep the test simple.
+template<typename T>
```
Comment (janden) [inline test/cuda/cufinufft_simple_test.cu]: This is quite general. Should this live in some utils file instead?

**Review by janden — CHANGES_REQUESTED** [review]
Looks great! Sorry it took me so long. Just a few questions.

File: src/cuda/c_interface.cpp (hunk 11 above)
Comment (DiamonDinoia) [reply src/cuda/c_interface.cpp:216]: setpts/execute return void and throw finufft::exception on failure; guru13/guru2 wrap simple_guru in safe_finufft_call, which catches and maps them to error codes. So they are handled

File: src/cuda/c_interface.cpp (hunk 12 above)
Comment (DiamonDinoia) [reply src/cuda/c_interface.cpp:225]: Yes — to preserve the public ABI's const on the input array (cj for types 1/3, fk for type 2) while simple_guru takes both non-const (the plan API does)

File: src/cuda/c_interface.cpp (hunk 13 above)
Comment (DiamonDinoia) [reply src/cuda/c_interface.cpp:230]: It runs the callable in a try/catch and converts any thrown finufft::exception into the matching integer error code for the C ABI (defined in finufft_common/safe_call.h). Documented in the same new comment.

File: test/cuda/cufinufft_simple_test.cu (hunk 14 above)
Comment (DiamonDinoia) [reply test/cuda/cufinufft_simple_test.cu]: Whoops, I asked claude to wrap up the test for me and it made up this instead of using the l2 like all other tests. Deleted this entirely

File: src/cuda/c_interface.cpp (hunk 11 above)
Comment (janden) [reply src/cuda/c_interface.cpp:216]: Ah of course. Was looking at the C interface here. Good.

File: src/cuda/c_interface.cpp (hunk 12 above)
Comment (janden) [reply src/cuda/c_interface.cpp:225]: Ok.

File: test/cuda/cufinufft_simple_test.cu (hunk 14 above)
Comment (janden) [reply test/cuda/cufinufft_simple_test.cu]: Oh ok. Very good!

**Review by janden — APPROVED** [review]
LGTM!
